cpp/src/parquet/encryption/properties_driven_crypto_factory.h

```diff
@@ -0,0 +1,210 @@
+
+    /// Lifetime of cached entities (key encryption keys, local wrapping keys, KMS client
+    /// objects).
+    /// The default is 600 (10 minutes).
+    Builder* cache_lifetime_seconds(uint64_t cache_lifetime_seconds);
+
+    /// Store key material inside Parquet file footers; this mode doesn’t produce
+    /// additional files. By default, true. If set to false, key material is stored in
+    /// separate files in the same folder, which enables key rotation for immutable
+    /// Parquet files.
+    Builder* internal_key_material(bool internal_key_material);
+
+    /// Length of data encryption keys (DEKs), randomly generated by parquet key
+    /// management tools. Can be 128, 192 or 256 bits.
+    /// The default is 128 bits.
+    Builder* data_key_length_bits(int32_t data_key_length_bits);
+
+    std::shared_ptr<EncryptionConfiguration> build();
+
+   private:
+    std::string footer_key_;
+    std::string column_keys_;
+    ParquetCipher::type encryption_algorithm_;
+    bool plaintext_footer_;
+    bool double_wrapping_;
+    uint64_t cache_lifetime_seconds_;
+    bool internal_key_material_;
+    bool uniform_encryption_;
+    int32_t data_key_length_bits_;
+  };
+
+  const std::string& footer_key() const { return footer_key_; }
+  const std::string& column_keys() const { return column_keys_; }
+  ParquetCipher::type encryption_algorithm() const { return encryption_algorithm_; }
+  bool plaintext_footer() const { return plaintext_footer_; }
+  bool double_wrapping() const { return double_wrapping_; }
+  uint64_t cache_lifetime_seconds() const { return cache_lifetime_seconds_; }
+  bool internal_key_material() const { return internal_key_material_; }
+  bool uniform_encryption() const { return uniform_encryption_; }
+  int32_t data_key_length_bits() const { return data_key_length_bits_; }
+
+ private:
+  EncryptionConfiguration(const std::string& footer_key, const std::string& column_keys,
+                          ParquetCipher::type encryption_algorithm, bool plaintext_footer,
+                          bool double_wrapping, uint64_t cache_lifetime_seconds,
+                          bool internal_key_material, bool uniform_encryption,
+                          int32_t data_key_length_bits)
+      : footer_key_(footer_key),
+        column_keys_(column_keys),
+        encryption_algorithm_(encryption_algorithm),
+        plaintext_footer_(plaintext_footer),
+        double_wrapping_(double_wrapping),
+        cache_lifetime_seconds_(cache_lifetime_seconds),
+        internal_key_material_(internal_key_material),
+        uniform_encryption_(uniform_encryption),
+        data_key_length_bits_(data_key_length_bits) {}
+
+  std::string footer_key_;
+  std::string column_keys_;
+  ParquetCipher::type encryption_algorithm_;
+  bool plaintext_footer_;
+  bool double_wrapping_;
+  uint64_t cache_lifetime_seconds_;
+  bool internal_key_material_;
+  bool uniform_encryption_;
+  int32_t data_key_length_bits_;
```

Comment:
So you have ten trivial accessors, plus a constructor with ten explicit arguments, plus a `Builder` subclass with ten setter methods... I don't understand the point of this. `EncryptionConfiguration` can be a simple struct with public attributes...

Reply:
I followed the same the way we write classes in https://github.com/apache/arrow/blob/master/cpp/src/parquet/encryption.h. I thought that were the way we should implement configuration class in arrow. No?

Reply:
Hmm... unfortunately, we don't have a uniform convention (especially as the Parquet C++ codebase used to live its own life), but in the future I think we'd rather use simple structs such as those:
https://github.com/apache/arrow/blob/master/cpp/src/arrow/csv/options.h

I'll turn the question on its head: the Builder pattern involves a lot of boilerplate and maintenance, so what does it bring compared to a simple struct?

Reply:
Actually I don't see any advantage to use `Builder` pattern compared to a simple struct in this case. I would love to move to a simple struct then.

Reply:
+1 from me. I mean, the builder pattern guarantees the object will be immutable; while with a struct it's easier for a developer to overlook unintended changes after the object is created. But if Arrow / parquet-cpp doesn't use the builder pattern, no reason to start using it now. This is one of the places where we already have a difference between Java and C++ encryption APIs - due to the substantial language differences.

Reply:
It's easy to make an attribute `const` if you want to avoid mutations.

Reply:
if this allows a user to set some attributes upon struct "constructions", while the rest of attributes will be automatically set to the default values - then this seems to be as good as builder's immutability stuff. There is an additional builder feature that allows to verify that mandatory parameters are set by the user - but that can be also checked by consumers of the struct. In any case, using a struct sounds good.